from snakemake.io import glob_wildcards

OUTPUT = config["output"]
PANGENOME_PATH = f"{OUTPUT}/pangenome"
TREE = config["tree_type"]
RULE_TYPE = config["rule_type"]

#pre-determined entries for rule all depending on which snakemake call it is
RULE_ALL_DICT = {
    "isolate": [f"{OUTPUT}/isolate_prokka/isolate.gff", f"{OUTPUT}/gtdbtk/gtdbtk.bac120.summary.tsv"],
    "gtotree": [f"{OUTPUT}/{TREE}_tree/{TREE}_tree.tre"]
}

#create wildcards to pull the unzipped .fa accessions into prokka rule to run in parallel
if RULE_TYPE == "pangenome":
    SAMPLES, = glob_wildcards(f"{PANGENOME_PATH}/fastas/{{sample}}.fa")
    RULE_ALL = [f"{PANGENOME_PATH}/roary/gene_presence_absence.csv"]
    include: "rules/multi_prokka.smk"
    # include: "rules/gffs.smk"
    include: "rules/roary.smk"

#select the rule depending on what config file variable is set as
else: 
    RULE_ALL = RULE_ALL_DICT[RULE_TYPE]

include: "rules/trim_reads.smk"
include: "rules/spades.smk"
include: "rules/prokka.smk"
include: "rules/gtdbtk.smk"
include: "rules/gtotree.smk"

rule all:
    input:
        RULE_ALL